src/libraries/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs

```diff
@@ -3936,6 +3936,150 @@ public static void DCS_SampleICollectionTExplicitWithoutDC()
         TestObjectWithDifferentPayload(value, netcorePayload, desktopPayload);
     }
 
+    [Fact]
+    public static void DCS_MemoryStream_Serialize_UsesBuiltInAdapter()
+    {
+        MemoryStream ms = new MemoryStream(new byte[] { 1, 2, 3, 4, 5 });
+        SerializeObjectAndValidate(ms, new byte[] { 1, 2, 3, 4, 5 }, 0, false);
+
+        ms = new MemoryStream(new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 }, 2, 5, true, true); // partial buffer
+        SerializeObjectAndValidate(ms, new byte[] { 3, 4, 5, 6, 7 }, 0, true);
+
+        ms = new MemoryStream(new byte[] { 1, 2, 3, 4, 5 }, 0, 4, writable: false, publiclyVisible: true) { Position = 2 };
+        SerializeObjectAndValidate(ms, new byte[] { 1, 2, 3, 4 }, 2, true);
+
+        ms = new MemoryStream(new byte[] { 1, 2, 3, 4, 5 }, 0, 4, writable: false, publiclyVisible: false) { Position = 4 };
+        SerializeObjectAndValidate(ms, new byte[] { 1, 2, 3, 4 }, 4, false);
+
+        static void SerializeObjectAndValidate(MemoryStream obj, byte[] expectedData, int expectedPosition, bool expectedExposable)
+        {
+            // First, inspect the XML manually, since we want to validate that the structure
+            // matches what we expect for the Full Framework.
+
+            StringWriter stringWriter = new StringWriter();
+            new DataContractSerializer(typeof(MemoryStream)).WriteObject(new XmlTextWriter(stringWriter), obj);
+            XElement root = XDocument.Parse(stringWriter.ToString()).Root;
+
+            XNamespace xns = "http://schemas.datacontract.org/2004/07/System.IO";
+
+            Assert.Equal(Convert.ToBase64String(expectedData), root.Element(xns.GetName("_buffer")).Value);
+            Assert.Equal(XmlConvert.ToString(0), root.Element(xns.GetName("_origin")).Value);
```

Comment:
We normally embed a known serialization of the class and compare the generated serialization with it. This code doesn't correctly validate the generated XML as DCS cares about ordering and this validation code will pass even if things are in the wrong order. If DCS expects the order A,B,C and the XML has elements ordered as B,A,C, the value of A won't get set. When it sees B, it goes through the list of potential consuming fields and matches for B. When it sees A, it starts at B in the list and doesn't find a consuming data member. This has a reasonable chance of currently being broken as DCS serializes alphabetically when using DataMember. You probably need to add the Order property on the attributes.